controllers/utils.go

```diff
@@ -55,3 +59,36 @@ func serviceDNSAddress(svc *corev1.Service) string {
 	// value for how this can be changed for a cluster.
 	return fmt.Sprintf("%s.%s.svc", svc.Name, svc.Namespace)
 }
+
+func addFinalizerIfNeeded(ctx context.Context, client client.Client, obj client.Object) error {
+	finalizer := deletionFinalizer(obj.GetObjectKind().GroupVersionKind().Kind)
+	if obj.GetDeletionTimestamp().IsZero() && !controllerutil.ContainsFinalizer(obj, finalizer) {
+		controllerutil.AddFinalizer(obj, finalizer)
+		if err := client.Update(ctx, obj); err != nil {
+			return fmt.Errorf("failed to add deletionFinalizer: %w", err)
+		}
+	}
+	return nil
+}
+
+func removeFinalizer(ctx context.Context, client client.Client, obj client.Object) error {
+	finalizer := deletionFinalizer(obj.GetObjectKind().GroupVersionKind().Kind)
+	controllerutil.RemoveFinalizer(obj, finalizer)
+	if err := client.Update(ctx, obj); err != nil {
+		return fmt.Errorf("failed to delete finalizer: %w", err)
+	}
+	return nil
+}
+
+// deletionFinalizer returns generated deletion finalizer
+// finalizers follow the format of deletion.finalizers.kind-plural-form.rabbitmq.com
+// for example: deletion.finalizers.bindings.rabbitmq.com and deletion.finalizers.policies.rabbitmq.com
```

Comment:
Could these just all be the same finalizer name? Would that be a problem?

Reply:
For a new operator, yes there won't be a problem. However for topology operator which already has previous versions, we need to make sure that finalizers stay the same to maintain compatibility with previous versions.